src/GenSymIO.chpl

```diff
@@ -1928,146 +1949,101 @@ module GenSymIO {
+     * Adjusts for the shuffling of a leading char sequence to the previous locale by 
+     * slicing leading chars that compose a string started in the previous locale and 
+     * returning a new char array.
      */
-    private proc sliceToValuesAndSegments(rawChars) throws {
-        var charList: list(uint(8), parSafe=false);
-        var indices: list(int, parSafe=false);
-
-        //initialize segments with index to first char in values
-        indices.append(0);
-        
-        for (value, i) in zip(rawChars, 0..rawChars.size-1) do {
-            charList.append(value:uint(8));
-            /*
-             * If the char is the null uint(8) char, check to see if it is the 
-             * last char. If not, added to the indices. If it is the last char,  
-             * don't add, because it is the correct ending char for a Strings 
-             * values array to be written to a locale.
-             */ 
-            if value == NULL_STRINGS_VALUE && i < rawChars.size-1 {
-                indices.append(i+1);
-            }
-        }
-
-        return (charList, indices);
-    }
+    private proc adjustCharArrayForLeadingSlice(sliceIndex, charArray, last, idx) throws { 
+        return charArray[sliceIndex..last]; 
+    }    
 
     /*
-     * Adjusts for the shuffling of a leading char sequence to the 
-     * previous locale by (1) slicing leading chars that compose
-     * a string started in the previous locale and returning (1) 
-     * a new values list that composes all of the strings that start
-     * in the current locale and (2) a new segments list that 
-     * corresponds to the new values list
+     * Adjusts for the left shuffle of the leading char sequence from the current locale
+     * to the previous locale by returning a slice containing chars from the shuffleLeftIndex
+     * to the end of the charList.
      */
-    private proc adjustForLeadingSlice(sliceIndex : int,
-                                   charList : list(uint(8))) throws {
-        var valuesList: list(uint(8), parSafe=false);
-        var indices: list(int);
-        var i: int = 0;
-        indices.append(0);   
-        var segmentsBound = charList.size - sliceIndex - 1;
-
-        for value in charList(sliceIndex..charList.size-1)  {
-            valuesList.append(value:uint(8));
-            
-            /*
-             * If the value is a null char and is not the last char
-             * in the list, then it is a segment use to delimit
-             * strings within the corresponding values array.
-             */
-            if value == NULL_STRINGS_VALUE && i < segmentsBound {
-                indices.append(i+1);
-            }
-            i+=1;
-        }
-
-        return (valuesList,indices);
+    private proc adjustForLeftShuffle(shuffleLeftIndex: int, charList, idx: int) throws {
+        return charList[shuffleLeftIndex..charList.size-1];
```

Comment:
One minor thing I noticed, the following are all basically doing the same thing, slicing an array/list at a given start/end points.

- adjustCharArrayForLeadingSlice
- adjustForLeftShuffle
- adjustForRightShuffle

In theory they could be reduced to a single genericSlicing function i.e.
```
private proc genericSlice(thingToSlice, startIndex, endIndex) throws {
    return thingToSlice[startIndex..endIndex];
}

// and the corresponding calls would be changed
// ~1354
- charList = new list(adjustForLeftShuffle(shuffleLeftIndex, charList, idx);
+ charList = new list(genericSlice(charList, shuffleLeftIndex, charList.size -1));

// ~1372
- charList = new list(adjustForRightShuffle(shuffleRightIndex, charList, idx));
+ charList = new list(genericSlice(charList, 0, shuffleRightIndex));

// ~1481
- var leadingCharArray = adjustCharArrayForLeadingSlice(localLeadingSliceIndex, A.localSlice(locDom), locDom.last, idx);
+ var leadingCharArray = genericSlice(A.localSlice(locDom), localLeadingSliceIndex, locDom.last);
```
However, you could easily make an argument that having separately named procs improves the clarity of what is going on, so it's the old _6 vs a half-dozen_ idiom.

Reply:
Good call, agree a common function is better, will make that change